apps/src/lib/tools/jsinterpreter/JSInterpreter.js

```diff
@@ -0,0 +1,1096 @@
+                                     typeof this.lastCallbackRetVal !== 'undefined') {
+      this.onExecutionWarning.notifyObservers("Function passed to onEvent() " +
+                                              "has taken too long - the return value was ignored.");
+      if (!this.shouldRunAtMaxSpeed()) {
+        this.onExecutionWarning.notifyObservers("  (try moving the speed " +
+                                                "slider to its maximum value)");
+      }
+    }
+  }
+
+  /**
+   * Yield execution (causes executeInterpreter loop to break out if this is
+   * called by APIs called by interpreted code)
+   */
+  yield() {
+    this.yieldExecution = true;
+  }
+
+  /**
+   * Find a bpRow from the "stopped at breakpoint" array by matching the scope
+   *
+   * @param {!Object} scope to match from the list
+   * @param {number} [row] to match from the list - in addition to scope
+   */
+  findStoppedAtBreakpointRow(scope, row) {
+    for (let i = 0; i < this.stoppedAtBreakpointRows.length; i++) {
+      const bpRow = this.stoppedAtBreakpointRows[i];
+      if (bpRow.scope === scope) {
+        if (typeof row === 'undefined' || row === bpRow.row) {
+          return bpRow;
+        }
+      }
+    }
+  }
+
+  /**
+   * Replace a bpRow from the "stopped at breakpoint" array by matching
+   * the scope.
+   *
+   * If no rows are found matching the given scope, a new one is introduced.
+   *
+   * @param {!Object} scope to match from the list
+   * @param {!number} row to replace in the list.
+   * @throws {TypeError} when given an invalid row.
+   */
+  replaceStoppedAtBreakpointRowForScope(scope, row) {
+    if (typeof row !== 'number' || row < 0) {
+      throw new TypeError('Row ' + row + ' is not a valid row in user code.');
+    }
+
+    for (let i = 0; i < this.stoppedAtBreakpointRows.length; i++) {
+      const bpRow = this.stoppedAtBreakpointRows[i];
+      if (bpRow.scope === scope) {
+        // Update row number
+        bpRow.row = row;
+        return;
+      }
+    }
+    // Scope not found, insert new object in array:
+    this.stoppedAtBreakpointRows.unshift({
+      row: row,
+      scope: scope
+    });
+  }
+
+  /**
+   * Remove a bpRow from the "stopped at breakpoint" array by matching
+   * the scope.
+   *
+   * Does nothing if no rows are found matching the given scope.
+   *
+   * @param {!Object} scope to match from the list
+   */
+  removeStoppedAtBreakpointRowForScope(scope) {
+    for (let i = 0; i < this.stoppedAtBreakpointRows.length; i++) {
+      const bpRow = this.stoppedAtBreakpointRows[i];
+      if (bpRow.scope === scope) {
+        // Remove from array
+        this.stoppedAtBreakpointRows.splice(i, 1);
+        return;
+      }
+    }
+  }
+
+  /**
+   * Determines if the program is done executing.
+   *
+   * @return {boolean} true if program is complete (or an error has occurred).
+   */
+  isProgramDone() {
+    const topStackFrame = this.interpreter.peekStackFrame();
```

Comment:
Possible null reference: Should probably pull the `this.interpreter` null check in the expression below up into this assignment.